Wasm/Bytes.lean

```diff
@@ -36,36 +35,30 @@ def ttoi (x : Type') : UInt8 :=
   | .f 32 => 0x7d
   | .f 64 => 0x7c
 
-def copp [Append α] : α → α → α :=
-  fun a x => x ++ a
-
-def totalLength (bss : List ByteArray) : Nat :=
-  bss.foldl (fun acc x => acc + x.data.size) 0
-
 def lindex (bss : ByteArray) : ByteArray :=
   uLeb128 bss.data.size ++ bss
 
--- TODO: Refactor completely and support locals!
-def extractTypes (m : Module) : ByteArray :=
-  let sigs := m.func.map $ fun x =>
-    let params := x.params.map $ (b ∘ ttoi ∘ fun x => x.type)
-    -- TODO: check if we support > 255 params. Do the same for each length and size entries!
-    let header := b 0x60 ++ uLeb128 params.length
-    let res := params.foldl Append.append header
-    res ++ (match x.results with -- TODO: test multi-result functions
-    | List.nil => b 0x00
-    | ts => b 0x7b ++ uLeb128 ts.length ++ flatten (ts.map $ b ∘ ttoi)
-    )
-  sigs.foldl
-    Append.append $
-      b 0x01 ++ uLeb128 (1 + totalLength sigs) ++ uLeb128 sigs.length
+def mkVec (xs : List α) (xtobs : α → ByteArray) : ByteArray :=
```

Comment:
Sometimes we need to flatten with indecies.

```
def flattenWithIndecies : List ByteArray → ByteArray
  | [] => b0
  | bs => (bs.foldl (fun (acc, n) x => match x.data with
    | #[] => (acc, n)
    | _x => ((uLeb128 n ++ x ++ acc), n + 1)) (b0, 0)).1

-- Same as extractModIdentifier, but maps a list of functions into a length-prefixed wasm array.
def extractFuncIdentifiers (fs : List Func) : ByteArray :=
  let fbs :=
    flattenWithIndecies $ fs.map extractFuncIdentifier
  if fbs.size = 0 then
    b0
  else
    funcHeader ++ (lindex $ uLeb128 fs.length ++ fbs)
```

`mkVec` helps a lot, but I think that it's not generic. 

If only tthere was a generalisation for such folds. Let's read on and find out.